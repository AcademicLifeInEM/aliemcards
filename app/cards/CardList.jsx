import React, { PropTypes } from 'react';
import { Link } from 'react-router';

const CardList = ({ cards, categories, filter, filterhandler, filtervalue, editortools }) =>
  <div>
    { filter &&
      <select value={filtervalue} onChange={filterhandler}>
        <option value="">Filter by Category:</option>
        {categories.map(category => (
          <option key={category.id} value={category.id}>{category.name}</option>
        ))}
      </select>
    }
    <div className="card-list">
      {cards.map(card =>
        <div key={card.id} className="card-list__item">
          { editortools &&
            <span className="card-list__item-meta ">{ new Date(card.created).toLocaleDateString('en-US') }</span>
          }
          <Link to={`/cards/${card.id}`} className="card-list__item-title">{card.title}</Link>{' '}
          {card.categories.map(category =>
            <Link to={`/categories/${category.id}`} key={`${card.id}-${category.id}`} className="card-list__item-meta">{category.name.toLowerCase()}</Link>
          )}
        </div>
      )}
    </div>
  </div>;

CardList.propTypes = {
  cards: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    title: PropTypes.string,
    categories: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.string,
      name: PropTypes.string,
    })),
  })),
  categories: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
  })),
  editortools: PropTypes.bool,
  filter: PropTypes.bool,
  filtervalue: PropTypes.string,
  filterhandler: PropTypes.func,
};

export default CardList;
