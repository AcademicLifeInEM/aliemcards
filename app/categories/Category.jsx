import React, { PureComponent, PropTypes } from 'react';
import { post } from 'axios';

import CardList from '../cards/CardList';

export default class Category extends PureComponent {

  static propTypes = {
    params: PropTypes.shape({
      category: PropTypes.string,
    }).isRequired,
  }

  constructor(props) {
    super(props);
    this.getCategory = this.getCategory.bind(this);
    this.state = {
      cards: [],
      category: {},
    };
  }

  componentDidMount() {
    this.getCategory();
  }

  componentDidUpdate(prevProps) {
    if (this.props.params.category !== prevProps.params.category) this.getCategory();
  }

  getCategory() {
    post('/graphql', {
      query: `query CategoryAndCards($category: String!){
        cards(category: $category) {
          id
          title
          categories {
            id
            name
          }
        }
        category (id: $category) {
          id
          name
        }
      }`,
      variables: { category: this.props.params.category },
    })
    .then(res => {
      if (res.status !== 200) throw res.status;
      const { cards, category } = res.data.data;
      console.log(cards);
      this.setState({
        cards,
        category,
      });
    })
    .catch(err => console.error(`Error: API response status code = ${err}`));
  }

  render() {
    return (
      <div>
        <h1>{this.state.category.name}</h1>
        <CardList cards={this.state.cards} />
      </div>
    );
  }
}
