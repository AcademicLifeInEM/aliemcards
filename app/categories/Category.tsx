import * as React from 'react';
import { gql, graphql } from 'react-apollo';
import { Card as ICard, Category as ICategory } from '../../server/models/';
import CardList from '../cards/CardList';

export interface Data {
  cards: Array<Pick<ICard, 'id'|'title'|'categories'>>;
  category: Pick<ICategory, 'id'|'name'>;
  networkStatus: number;
}

interface Props {
  data: Data;
  match: {
    params: {
      category: string;
    };
  };
}

const currentCategoryAndCards = gql`
  query CategoryAndCards($category: String!) {
    cards(category: $category) {
      id
      title
      categories {
        id
        name
      }
    }
    category(id: $category) {
      id
      name
    }
  }
`;

const config = {
  options: ({ match }) => ({ variables: { category: match.params.category } }),
};

@graphql(currentCategoryAndCards, config)
export default class Category extends React.PureComponent<Props, {}> {
  render() {
    const { cards, category, networkStatus } = this.props.data;
    if (networkStatus === 6) {
      return <div>Polling!</div>;
    }
    if (networkStatus < 7) {
      return <div>Loading...</div>;
    }
    return (
      <div>
        <h1>{category.name}</h1>
        <CardList cards={cards || []} />
      </div>
    );
  }
}
