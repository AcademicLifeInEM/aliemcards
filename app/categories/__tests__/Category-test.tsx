jest.mock('react-apollo');
jest.mock('../../cards/CardList');

import * as React from 'react';
import * as renderer from 'react-test-renderer';
import CardList from '../../cards/CardList';
import Category, { Data } from '../Category';

(CardList as any).mockImplementation(() => <div>Mocked CardList</div>);

const stub: any = {};
const mockData = {
  category: {
    id: 'test-category',
    name: 'Test Category',
  },
  networkStatus: 1,
};

const setup = (networkStatus: number = 1, cards: any = []) => {
  const data: Data = {
    ...mockData,
    cards,
    networkStatus,
  };
  const snapshot = renderer.create(<Category data={data} match={stub} />);
  return {
    snapshot,
  };
};

describe('<Category />', () => {
  it('should render appropriately depending on networkStatus', () => {
    const { snapshot: loading } = setup();
    expect(loading.toJSON()).toMatchSnapshot();

    const { snapshot: polling } = setup(6);
    expect(polling.toJSON()).toMatchSnapshot();

    const { snapshot: withCards } = setup(7);
    expect(withCards.toJSON()).toMatchSnapshot();

    const { snapshot: withoutCards } = setup(7, null);
    expect(withoutCards.toJSON()).toMatchSnapshot();
  });
});
