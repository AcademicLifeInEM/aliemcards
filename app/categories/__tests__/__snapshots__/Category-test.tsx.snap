// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Category /> should render appropriately depending on networkStatus 1`] = `
<div>
  Loading...
</div>
`;

exports[`<Category /> should render appropriately depending on networkStatus 2`] = `
<div>
  Polling!
</div>
`;

exports[`<Category /> should render appropriately depending on networkStatus 3`] = `
<div>
  <h1>
    Test Category
  </h1>
  <div>
    Mocked CardList
  </div>
</div>
`;

exports[`<Category /> should render appropriately depending on networkStatus 4`] = `
<div>
  <h1>
    Test Category
  </h1>
  <div>
    Mocked CardList
  </div>
</div>
`;
